### PR TITLE
Simplify variant handling

### PR DESCRIPTION
**Pull request by twuttke**
Handling variants in the callbacks is awkward, and I am guessing nobody really needed that granularity so far. By dealing with it in the regex, it simplifies logic in preparation for Diversity Emoji, which is coming next. From now on, you won't even get a match or a callback if the emoji has been suppressed by a valid use of U+FE0E.

Instead of "icon", "variant", callbacks get the "rawText", which is the complete text of the Emoji, including any optional variant selector codes if they really need to parse them. Variants aren't always at the end of the Emoji match sequence in Diversity, Zero Width Joiners, or Keycaps, so we need to break this pattern of assuming they are. 

Callbacks also may get an iconId, when it is the resolved name of the png/svg without the variant, like "2764" or "1f1fa-1f1f8" The name should prevent any confusion with rawText.

Let me know what you think. I know this changes the API slightly, so I am open for discussion. All tests pass. :)

- Tom

### CODE REVIEW

**Comment by WebReflection** [comment]
I think it makes sense but this part:
> From now on, you won't even get a match or a callback if the emoji has been suppressed by a valid use of U+FE0E
was actually the reason I've put variants in the first place: backward, non emoji at all, devices.

I understand the standard wants FE0E to mean "don't change anything" but IIRC there were few cases where this was flagged as inappropriate because some device wasn't showing anything (mostly old browsers related gotcha).

Anyway, if you think this is going to improve diversity fine to me, I'm just worried somebody might have used the old API and if they use MaxCDN which doesn't provide versioning things might break all over.

Are you confident this change wont damage much?

**Comment by twuttke** [comment]
You have a valid point about it breaking a few live sites that might be making heavy use of the callbacks + variant handling and pulling live from the CDN since it isn't versioned. Is there a way to alert them to fix their site right away?  I suppose a day or two of broken Emojis might do it. The other way I can go is to pass an empty string for the callbacks in the variant spot for backward compat. I'd be surprised if this change had any significant problems with any live sites, though. One of the callbacks that took (icon, variant) now takes (rawText, iconId) If people had code written that would generate a src only if the second param was not U+FE0E, that would still work since the iconID will never equal that.

As for how many sites would benefit from being able to intercept U+FE0E sequences.... if there was an old browser that could not render U+2764,U+FE0E as plain text for example, giving them a callback for it to replace it with [smiley] might seem helpful, but think of all the other unicode chars that browser won't display properly that are not on our list of 800 or so "emojis."  I think if an author is using U+FE0E they are making a specific request that it be plain text. I measured at one point less than 5 Tweets per day used U+FE0E out of hundreds of millions, so I don't think this case needs a ton of attention.  >99.999% of Emoji instances use U+FE0F or no variant at all.

I'm inclined to go forward with this, and then do diversity over the weekend. 

Question - how do the minimized versions get created?  Do I need to do anything? I didn't see release instructions (maybe I didn't look hard enough)

**Comment by WebReflection** [comment]
the build process should be better described.

TL;DR the only file you need to change is the [twemoji-generator](https://github.com/twitter/twemoji/blob/gh-pages/twemoji-generator.js) one.

One you `./twemoji-generator.js` it [invokes at the end](https://github.com/twitter/twemoji/blob/gh-pages/twemoji-generator.js#L265-L269) the [dist file](https://github.com/twitter/twemoji/blob/gh-pages/twemoji-dist.js) which takes care of creating amd, npm, and min versions.

To be sure it works you need `uglifyjs` which is brought in via `npm install`.

To test everything run `npm test` and ideally try the `index.html` with some browser to see green-light.

Please let me know once you're confident everything is OK so I can merge this.

No need to ship a patched API, I'd rather call the file `twemoji2.js` to avoid any possible issue but then we have an upside-down problem, who's using the CDN won't be automagically updated with Unicode 8 and diversity related entries.

decisions ... decisions ...

**Comment by twuttke** [comment]
OK, I think I know what happened. I had an old version of node on this box, and it didn't spew any errors when the spawn child processes failed. Also, there is uglify, uglifyjs, both of which are valid npm installs that are NOT the same as uglify-js

**Comment by twuttke** [comment]
I'll let you decide if we should fork the whole thing to twemoji2.js or just provide two js files or what. Updating in place will probably be fine, but there are advantages to being clear v2 is a little different.

**Comment by WebReflection** [comment]
Hi Tom, the build process is messy and maybe this is the right time to simplify it.

Here what I've just pushed, a folder `2` with inside a folder `utils`.

Basically, since the way it works with MaxCDN here is that whatever we push in the repo will be replicated in their CDN, the major version `2` of this library will be reachable simply adding `/2/twemoji.min.js` instead of just `/twemoji.min.js` in the url.

This grants no problems with the current version and it gives us a new folder to release major version 2.

Tags also will follow same convention, so these will start as v2.0.0 as soon as your changes have been merged.

I have updated the `package.json` because it was pointing at the wrong version of phantomjs.

As soon as you'll manage to change the file within the `2/utils/generate` file in order to include your changes, I'll merge that and update the TravisCI job and current tests in order to point at the version 2.

I will write that version 1 is in maintenance mode and will describe changes in version 2 as soon as your files are in. As long as we don't change anything in the root folder we are free to "play" with changes and see how it works.

TL;DR please rebase your PR and change only the `2/utils/generate` file with your proposal so that your new PR will only affect the folder `2` and nothing else so we can all sleep well tonight :-)

Hope this works for you too, if you have better ideas let me know.

Thanks

**Comment by WebReflection** [comment]
Hi Tom,
  I've done even more. I've manually merged and double checked your changes and put back the test you removed with a test that verifies actually nothing happens as expected so we have 73 assertions in both versions.

I've configured the `package.json` to run in parallele both tests for version 1 and version 2, and the new structure to generate, create disttributed files and test them has been pushed to [the folder ./2](https://github.com/twitter/twemoji/tree/gh-pages/2)

At this point there's no need for you to do anything, just remember to edit `2/utils/generate` file for possible extra changes and that's it, we have a non breaking API change on a dedicated folder with a much cleaner way to test and build.

Let me know what you think about it.

Eventually we need to update the README deprecating version 1 and promoting version 2 as soon as we have in diversity.

At that point we can also update the preview page to include all diversity using the version 2.

Best Regards
